src/raftstore/store/util.rs

```diff
@@ -534,6 +583,85 @@ impl fmt::Debug for Lease {
     }
 }
 
+/// A remote lease, it can only be derived by `Lease`. It will be sent
+/// to the local read thread, so name it remote. If Lease expires, the remote must
+/// expire too.
+#[derive(Debug)]
+pub struct RemoteLease {
+    expired_time: Arc<AtomicU64>,
+    term: u64,
+}
+
+impl RemoteLease {
+    pub fn inspect(&self, ts: Option<Timespec>) -> LeaseState {
+        let expired_time = self.expired_time.load(AtomicOrdering::Acquire);
+        if ts.unwrap_or_else(monotonic_raw_now) < u64_to_timespec(expired_time) {
+            LeaseState::Valid
+        } else {
+            LeaseState::Expired
+        }
+    }
+
+    fn renew(&self, bound: Timespec) {
+        self.expired_time
+            .store(timespec_to_u64(bound), AtomicOrdering::Release);
+    }
+
+    fn expire(&self) {
+        self.expired_time.store(0, AtomicOrdering::Release);
+    }
+
+    pub fn term(&self) -> u64 {
+        self.term
+    }
+}
+
+/// Convert Timespec to u64. It's millisecond precision and
+/// covers a range of about 571232829 years in total.
+///
+/// # Panics
+///
+/// If Timespecs have negative sec.
+#[inline]
+fn timespec_to_u64(ts: Timespec) -> u64 {
+    // > Darwin's and Linux's struct timespec functions handle pre-
+    // > epoch timestamps using a "two steps back, one step forward" representation,
+    // > though the man pages do not actually document this. For example, the time
+    // > -1.2 seconds before the epoch is represented by `Timespec { sec: -2_i64,
+    // > nsec: 800_000_000 }`.
+    //
+    // Quote from crate time,
+    //   https://github.com/rust-lang-deprecated/time/blob/
+    //   e313afbd9aad2ba7035a23754b5d47105988789d/src/lib.rs#L77
+    assert!(ts.sec >= 0 && ts.sec < (1i64 << (64 - TIMESPEC_SEC_SHIFT)));
+    assert!(ts.nsec >= 0);
+
+    // Round down to millisecond precision.
+    let ms = ts.nsec >> TIMESPEC_NSEC_SHIFT;
+    let sec = ts.sec << TIMESPEC_SEC_SHIFT;
+    sec as u64 | ms as u64
+}
+
+const NSEC_PER_MSEC: i32 = 1_000_000;
```

Comment:
Seems these consts are in nowhere in particular. Maybe they can go above all their uses (not just some), or at the top of the file?